pkg/command/upgrade/upgrade.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/singularity-data/risingwave-operator/pkg/command/util"
+)
+
+type Options struct {
+	*cmdcontext.BasicOptions
+
+	version string
+}
+
+const (
+	LongDesc = `
+Upgrade a risingwave instance to a specified version.
+`
+	Example = `  # Upgrade risingwave named example-rw to the latest version.
+  kubectl rw upgrade example-rw
+
+  # Upgrade risingwave named example-rw in namespace foo to the nightly version.
+  kubectl rw upgrade example-rw -n foo -v nightly
+`
+)
+
+func NewCommand(ctx *cmdcontext.RWContext, streams genericclioptions.IOStreams) *cobra.Command {
+	o := Options{
+		BasicOptions: cmdcontext.NewBasicOptions(streams),
+	}
+
+	cmd := &cobra.Command{
+		Use:     "upgrade",
+		Short:   "Upgrade the risingwave instance to a specified version",
+		Long:    LongDesc,
+		Example: Example,
+		Run: func(cmd *cobra.Command, args []string) {
+			util.CheckErr(o.Complete(ctx, cmd, args))
+			util.ExitOnErr(o.Validate(ctx, cmd, args))
+			util.CheckErr(o.Run(ctx, cmd, args))
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.version, "version", "v", "latest", "The version to upgrade to. If not specified, the latest version will be used.")
+
+	return cmd
+}
+
+func (o *Options) Validate(ctx *cmdcontext.RWContext, cmd *cobra.Command, args []string) error {
+	err := o.verifyTag()
+	if err != nil {
+		return err
+	}
+
+	// parse current version
+	rw, err := o.GetRwInstance(ctx)
+	if err != nil {
+		return err
+	}
+
+	if rw.Spec.Global.Image == fmt.Sprintf("ghcr.io/singularity-data/risingwave:%s", o.version) {
+		return fmt.Errorf("%s is already the current version", o.version)
+	}
+
+	return nil
+}
+
+func (o *Options) Run(ctx *cmdcontext.RWContext, cmd *cobra.Command, args []string) error {
+	rw, err := o.GetRwInstance(ctx)
+	if err != nil {
+		return err
+	}
+
+	o.updateTag(rw)
+
+	err = ctx.Client().Update(context.Background(), rw)
+	if err != nil {
+		return fmt.Errorf("failed to update instance, %v", err)
+	}
+
+	return nil
+}
+
+// compare it to the current version.
+func (o *Options) updateTag(rw *v1alpha1.RisingWave) {
+	image := fmt.Sprintf("ghcr.io/singularity-data/risingwave:%s", o.version)
+
+	rw.Spec.Global.Image = image
+	for i := range rw.Spec.Components.Compactor.Groups {
+		rw.Spec.Components.Compactor.Groups[i].Image = image
+	}
+	for i := range rw.Spec.Components.Compute.Groups {
+		rw.Spec.Components.Compute.Groups[i].Image = image
+	}
+	for i := range rw.Spec.Components.Frontend.Groups {
+		rw.Spec.Components.Frontend.Groups[i].Image = image
+	}
+	for i := range rw.Spec.Components.Meta.Groups {
+		rw.Spec.Components.Meta.Groups[i].Image = image
+	}
+}
+
+// uses the github api to get the latest version.
+// https://github.com/orgs/community/discussions/26279#discussioncomment-3251171
+func (o *Options) verifyTag() error {
+	// curl https://ghcr.io/token\?scope\="repository:singularity-data/risingwave:pull"
+	token := "djE6c2luZ3VsYXJpdHktZGF0YS9yaXNpbmd3YXZlOjE2NjAyMDg1MTU1NDgyODMwMDU="
```

Comment:
Is this token granted with limited privileges? Otherwise this could be a security issue.

Reply:
ghrc requires a token for this operation even if it is a public package (which it is in this case). This token is a fake NOOP one which should have no privileges.

Reply:
[this](https://github.com/orgs/community/discussions/26279#discussioncomment-3251171) thread provides more information

Reply:
oh, i see. Thanks for the explanation.